tools/update-rule-type-headers.js

```diff
@@ -0,0 +1,324 @@
+
+    return newTSDoc;
+}
+
+/**
+ * Updates rule header comments in a `.d.ts` file.
+ * @param {string} ruleTypeFile Pathname of the `.d.ts` file.
+ * @param {Set<string>} consideredRuleIds The names of the rules to be considered for the current run.
+ * @returns {Promise<Iterable<string>>} The names of the rules found in the `.d.ts` file.
+ */
+async function updateTypeDeclaration(ruleTypeFile, consideredRuleIds) {
+    const sourceText = await readFile(ruleTypeFile, "utf-8");
+    const tsDocRangeMap = getTSDocRangeMap(sourceText, consideredRuleIds);
+    const chunks = [];
+    let lastPos = 0;
+
+    for (const [ruleId, [tsDocStart, tsDocEnd]] of tsDocRangeMap) {
+        const textBeforeTSDoc = sourceText.slice(lastPos, tsDocStart);
+        const currentTSDoc = sourceText.slice(tsDocStart, tsDocEnd);
+        const newTSDoc = createTSDoc(ruleId, currentTSDoc);
+
+        chunks.push(textBeforeTSDoc, newTSDoc);
+        if (sourceText[tsDocEnd] !== "\n") {
+            chunks.push("\n    ");
+        }
+        lastPos = tsDocEnd;
+    }
+    chunks.push(sourceText.slice(Math.max(0, lastPos)));
+    const newText = chunks.join("");
+
+    await writeFile(ruleTypeFile, newText);
+    return tsDocRangeMap.keys();
+}
+
+//-----------------------------------------------------------------------------
+// Main
+//-----------------------------------------------------------------------------
+
+(async () => {
+    const consideredRuleIds = getConsideredRuleIds();
+    const ruleTypeDir = join(__dirname, "../lib/types/rules");
+    const fileNames = (await readdir(ruleTypeDir)).filter(fileName => extname(fileName) === ".ts");
+    const typedRuleIds = new Set();
+    const repeatedRuleIds = new Set();
+    const untypedRuleIds = [];
+
+    console.log(`Considering ${consideredRuleIds.size} rule(s).`);
+    await Promise.all(
+        fileNames.map(
+            async fileName => {
+                const ruleIds = await updateTypeDeclaration(join(ruleTypeDir, fileName), consideredRuleIds);
+
+                for (const ruleId of ruleIds) {
+                    if (typedRuleIds.has(ruleId)) {
+                        repeatedRuleIds.add(ruleId);
+                    } else {
+                        typedRuleIds.add(ruleId);
+                    }
+                }
+            }
+        )
+    );
+    for (const ruleId of consideredRuleIds) {
+        if (!typedRuleIds.has(ruleId)) {
+            untypedRuleIds.push(ruleId);
+        }
+    }
+    if (repeatedRuleIds.size) {
+        console.warn(
+            "The following rules have multiple type definition:%s",
+            [...repeatedRuleIds].map(ruleId => `\n* ${ruleId}`).join("")
+        );
+        process.exitCode = 1;
+    }
+    if (untypedRuleIds.length) {
+        console.warn(
+            "The following rules have no type definition:%s",
+            untypedRuleIds.map(ruleId => `\n* ${ruleId}`).join("")
+        );
+        process.exitCode = 1;
+    }
```

Comment:
Do we have CI checks that would catch these problems? If not, what do you think about running this tool in CI just to check the exit code, in order to catch the problems before releases (as that wouldn't be a good time for fixing the types)?

Reply:
Perhaps along (right before) running `npm run test:types` for the `eslint` package:

https://github.com/eslint/eslint/blob/b04b84bc17d4aaaea1326cb08196593624db02a2/.github/workflows/ci.yml#L147-L149

Reply:
> Do we have CI checks that would catch these problems? If not, what do you think about running this tool in CI just to check the exit code, in order to catch the problems before releases (as that wouldn't be a good time for fixing the types)?

There are currently no checks for these problems, which is why some rules are missing type definitions. I think it's a good idea to run the script during CI to catch problems before release. We could also throw an error to fail the build when the `.d.ts` files are modified by the script during CI (meaning that lint-staged didn't run), or alternatively we could add a `--check` flag to the script to throw an error instead of saving the `.d.ts` file if there are any changes.

Reply:
The `--check` flag sounds good to me. @nzakas what do you think?

Reply:
Sounds good to me. :+1:

Reply:
Thanks for the suggestions. I've added a `--check` option to the tool and a `lint:rule-types` script to package.json that runs the tool in check mode. And I've updated the "Verify Files" job in CI to run this check along with the other lint steps. We could also run the check during "Test types for eslint" as initially suggested.